### PR TITLE
istioctl upgrade: check previous version of Istio is installed with revision.

### LINKED ISSUE #28566: istioctl upgrade: failed to upgrade if previous version is installed with --revision
```
[istio-1.7.3] $ istioctl install -r 1-7-3

[1.8.0-alpha.2] $ istioctl upgrade
```

Goes into an infinite loop and the upgrade is never finishedd.
```
$ istioctl upgrade
2020-11-04T07:15:07.053176Z     info    proto: tag has too few fields: "-"
Control Plane - ingressgateway pod - istio-ingressgateway-5cf98b4d7d-vrqxc - version: 1.7.3
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3

Upgrade version check passed: 1.7.3 -> 1.8.0.

Confirm to proceed [y/N]? y
✔ Istio core installed                                                                                                                                             
✔ Istiod installed                                                                                                                                                 
✔ Ingress gateways installed                                                                                                                                       
✔ Installation complete                                                                                                                                            ..........
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3 does not match the target version 1.8.0
..........
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3 does not match the target version 1.8.0
..........
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3 does not match the target version 1.8.0
....
```

If you install Istio without the `--revision` flag then the upgrade went smoothly.

```
[istio-1.7.3] $ istioctl install
[1.8.0-alpha.2] $ istioctl upgrade

Confirm to proceed [y/N]? y
✔ Istio core installed                                                                                                                                             
✔ Istiod installed                                                                                                                                                 
✔ Ingress gateways installed                                                                                                                                       
✔ Installation complete                                                                                                                                            ..........
Upgrade rollout completed. All Istio control plane pods are running on the target version.

Control Plane - ingressgateway pod - istio-ingressgateway-7856bd4877-766k6 - version: 1.8.0
Control Plane - pilot pod - istiod-6ff7599b89-cpwpr - version: 1.8.0

Success. Now the Istio control plane is running at version 1.8.0.

To upgrade the Istio data plane, you will need to re-inject it.
If you’re using automatic sidecar injection, you can upgrade the sidecar by doing a rolling update for all the pods:
    kubectl rollout restart deployment --namespace <namespace with auto injection>
If you’re using manual injection, you can upgrade the sidecar by executing:
    kubectl apply -f < (istioctl kube-inject -f <original application deployment yaml>
```

### PR DESCRIPTION
**Pull request by shamsher31**
To fix https://github.com/istio/istio/issues/28566

Based on https://github.com/istio/istio/issues/28566#issuecomment-721937551

Before:
```
[istio-1.7.3] $ istioctl install -r 1-7-3

[1.8.0-alpha.2] $ istioctl upgrade
```
Goes into an infinite loop and the upgrade is never finished.
```
$ istioctl upgrade
2020-11-04T07:15:07.053176Z     info    proto: tag has too few fields: "-"
Control Plane - ingressgateway pod - istio-ingressgateway-5cf98b4d7d-vrqxc - version: 1.7.3
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3

Upgrade version check passed: 1.7.3 -> 1.8.0.

Confirm to proceed [y/N]? y
✔ Istio core installed                                                                                                                                             
✔ Istiod installed                                                                                                                                                 
✔ Ingress gateways installed                                                                                                                                       
✔ Installation complete                                                                                                                                            ..........
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3 does not match the target version 1.8.0
..........
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3 does not match the target version 1.8.0
..........
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3 does not match the target version 1.8.0
.
.
.
.
.
Control Plane - istiod pod - istiod-1-7-3-b5c499b7b-xzmvp - version: 1.7.3 does not match the target version 1.8.0
........
```

After:
```
$ ./out/linux_amd64/istioctl upgrade -d manifests/
2020-11-19T10:04:42.916754Z     info    proto: tag has too few fields: "-"
Control Plane - ingressgateway pod - istio-ingressgateway-69cccb7b54-sltvf - version: 1.8.0
Control Plane - istiod pod - istiod-1-8-0-7d9947f98b-z52zs - version: 1.8.0

2020-11-19T10:04:42.935391Z     info    Error: can not upgrade because the previous version of Istio is installed with revision.
Use canary upgrades instead: https://istio.io/v1.8/docs/setup/upgrade/canary

Error: can not upgrade because the previous version of Istio is installed with revision.
Use canary upgrades instead: https://istio.io/v1.8/docs/setup/upgrade/canary

```

Related:
https://github.com/istio/istio.io/pull/8447

### CODE REVIEW

File: operator/cmd/mesh/upgrade.go
Comment (ostromart) [inline operator/cmd/mesh/upgrade.go:173]: shouldn't we continue if --revision is set?

Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go:173]: > But regardless, the command itself should be able to check the cluster to see if there's a revision label on istiod and fail with an error in that case rather than proceeding as it does. https://github.com/istio/istio/issues/28566#issuecomment-721937551

If the revision is set other than `default`, then the upgrade goes into the infinite loop.

Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go:173]: Maybe we can add an additional info in the error message saying `Prefer canary upgrades`?
https://preliminary.istio.io/latest/docs/setup/upgrade/canary/

Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go:178]: @ostromart PTAL.

Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go:178]: The output will look like
```
Error: can not upgrade because the previous version of Istio is installed with revision.
Use canary upgrades instead: https://istio.io/v1.9/docs/setup/upgrade/canary
```

**Comment by shamsher31** [comment]
@ostromart did you got a chance to look into this.

**Comment by shamsher31** [comment]
@esnible Do you mind taking a look and provide your feedback on this?

**Review by richardwxn — COMMENTED** [review]
instead of preventing the users from doing an upgrade from existing installation with revision, it might be better to fix the bugs that cause the infinite loop. Looked at the code, it seems like there are issues of the logic that check Istio versions

**Comment by shamsher31** [comment]
> istioctl upgrade is for in-place upgrade and not compatible with revisions. So the first thing would be to change the docs to mention that. But regardless, the command itself should be able to check the cluster to see if there's a revision label on istiod and fail with an error in that case rather than proceeding as it does. https://github.com/istio/istio/issues/28566#issuecomment-721937551

The current implementation is based on @ostromart suggestion. 

> Looked at the code, it seems like there are issues of the logic that check Istio versions

@richardwxn Let me know what issue you found and I will try to work on it. Thanks.

**Comment by shamsher31** [comment]
cc: @ostromart @richardwxn

**Comment by esnible** [comment]
Is this still happening?  If so, what I can do to help move this forward?

**Comment by shamsher31** [comment]
I'm waiting for a review from @istio/wg-environments-maintainers 

Docs are updated for upgrade message. https://github.com/istio/istio.io/pull/8447

File: operator/cmd/mesh/upgrade.go
Comment (elfinhe) [inline operator/cmd/mesh/upgrade.go:173]: -> "failed to fetch the existing istiod pod for checking revision, error: %v"

File: operator/cmd/mesh/upgrade.go
Comment (elfinhe) [inline operator/cmd/mesh/upgrade.go:172]: we can add `&& !args.force` to let force flag work

File: operator/cmd/mesh/upgrade.go
Comment (elfinhe) [inline operator/cmd/mesh/upgrade.go:179]: instead we can use a list of errors?

File: operator/cmd/mesh/upgrade.go
Comment (elfinhe) [inline operator/cmd/mesh/upgrade.go:178]: There is a slight difference between "there is a revisioned istiod (and a non-revisioned istiod)" and "all istiod are revisioned". I think we only need to stop the latter.

But we still need to let force flag to skip the check.

**Comment by esnible** [comment]
I have no problems with this, but I haven't actually pulled it and tried it.  The UX workgroup doesn't have authority over the changed methods so I can't approve it.  It would be good if anything related to `upgrade` is merged before the 1.9 alpha is built so that we get a lot of eyes on it during testing.

Please @istio/wg-environments-maintainers help this along.

File: operator/cmd/mesh/upgrade.go
Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go]: Done.

File: operator/cmd/mesh/upgrade.go
Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go:178]: Right, and it only stops if the previous istio is installed with the `revision` flag. If the istio is installed without a revision flag (which means its default) then it proceeds with a successful upgrade.

File: operator/cmd/mesh/upgrade.go
Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go:179]: I don't want to change the semantics of the upgrade. The idea is to fail as early as possible and provide a useful message to the user for any upgrade related issues.

File: operator/cmd/mesh/upgrade.go
Comment (shamsher31) [reply operator/cmd/mesh/upgrade.go:172]: Added. Thanks.

**Comment by shamsher31** [comment]
cc: @elfinhe

**Comment by shamsher31** [comment]
@elfinhe PTAL.

**Review by elfinhe — COMMENTED** [review]
LGTM in general. 

But like richardwxn mentioned, this is not actually fixing the endless loop when a revision presents in istio-system. Maybe we can also add the revision check at L367.


A follow up question - should we enhance  `istioctl upgrade` to support in-place upgrade an istio with revision?

**Comment by shamsher31** [comment]
> A follow up question - should we enhance istioctl upgrade to support in-place upgrade an istio with revision?

Yes, I think we should look into that, but I will prefer to at least fix the issue for now so that user upgrading their Istio won't face this problem.

**Review by elfinhe — APPROVED** [review]

**Comment by shamsher31** [comment]
@linsun PTAL.

**Comment by shamsher31** [comment]
cc: @richardwxn @ostromart @nmittler @linsun

**Comment by shamsher31** [comment]
Ping @istio/wg-environments-maintainers

**Comment by shamsher31** [comment]
cc: @richardwxn @ostromart PTAL.

**Comment by howardjohn** [comment]
Coming back to try to clean up/move along old PRs.. is this still desired? `upgrade` command was removed, so this would need to be reworked, but seems like it could be used for `install` as well

**Pull request closed without merging**
